cmd/libsnap-confine-private/mount-opt.c

```diff
@@ -241,3 +243,61 @@ const char *sc_umount_cmd(char *buf, size_t buf_size, const char *target,
 
 	return buf;
 }
+
+void sc_do_mount(const char *source, const char *target,
+		 const char *fs_type, unsigned long mountflags,
+		 const void *data)
+{
+	char buf[10000];
+	const char *mount_cmd = NULL;
+
+	void ensure_mount_cmd() {
+		if (mount_cmd == NULL) {
+			mount_cmd = sc_mount_cmd(buf, sizeof buf, source,
+						 target, fs_type, mountflags,
+						 data);
+		}
+	}
+
+	if (sc_is_debug_enabled()) {
+		ensure_mount_cmd();
+		debug("performing operation: %s", mount_cmd);
+	}
+	if (sc_faulty("mount", NULL)
+	    || mount(source, target, fs_type, mountflags, data) < 0) {
+		// Save errno as ensure can clobber it.
+		int saved_errno = errno;
+		ensure_mount_cmd();
```

Comment:
This means that a failed mount goes through the whole string handling machinery of sc_mount_cmd() which I thought we agreed we would only do in debug mode, not production mode.